index.js

```diff
@@ -747,17 +747,25 @@
 
   //  Array$chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
   function Array$chainRec(f, x) {
-    var $todo = [x];
-    var $done = [];
-    while ($todo.length > 0) {
-      var xs = f(iterationNext, iterationDone, $todo.shift());
-      var $more = [];
-      for (var idx = 0; idx < xs.length; idx += 1) {
-        (xs[idx].done ? $done : $more).push(xs[idx].value);
+    var result = [];
+    var nil = {};
+    var xs = {head: x, tail: nil};
+    while (xs !== nil) {
+      var sx = nil;
+      f(iterationNext, iterationDone, xs.head).forEach(function(step) {
+        if (step.done) {
+          result.push(step.value);
+        } else {
+          sx = {head: step.value, tail: sx};
+        }
+      });
+      xs = xs.tail;
+      while (sx !== nil) {
+        xs = {head: sx.head, tail: xs};
```

Comment:
The $more was there so during iterations order was "respected". ie result of each "step" was prepended to $todo - conceptually like this `$more <> $todo`. this version using list conseptually does this `flip $more <> $todo`, if `flip` is fine, and we don't want to "respect" order then current version could be changed so apply is not needed:

```diff
//  Array$chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
function Array$chainRec(f, x) {
  var $todo = [x];
  var $done = [];
  while ($todo.length > 0) {
    var xs = f(iterationNext, iterationDone, $todo.shift());
-    var $more = [];
    for (var idx = 0; idx < xs.length; idx += 1) {
-      (xs[idx].done ? $done : $more).push(xs[idx].value);
+      (xs[idx].done ? $done : $todo).push(xs[idx].value);
    }
-    Array.prototype.unshift.apply($todo, $more);
  }
  return $done;
}```


but I think because of the laws  it's not allowed to reorder staff so we have to respect the order.

Tests are passing because the input to the laws i guess contains same values so order is not checked, if in tests we use ordered values then it would be more clear why we should respect the order.

for example this will pass in all versions as the flip is used when there is `next` returned during iteration

```js
Array.chainRec((next,done, x) => {
  return Array.from({length: x}, (_, k) => done(k+1)); 
}, [1,2,3,4,5,6])
```
should evaluate to `[1,1,2,1,2,3,1,2,3,4,1,2,3,4,5,1,2,3,4,5,6]`

but if stepping function returned next then current version will be broken i think, test using something like this:

```
Array.chainRec((next,done, x) => {
  return Array.from({length: x.v}, (_, k) => x.d ? done(k+1) : next({v:k+1, d:true })); 
}, [{v:2, d:false }
   ,{v:3, d:false }
   ]).join('')
```
I think this should evaluate to `112112123` if implementation is correct

Reply:
also old names more todo done are better i think than xs ys ...

Reply:
> old names more todo done are better i think than xs ys

You're right. I'll revert to the descriptive names. I named the linked lists `xs` and `sx` to hint that the latter is in reverse order, but this is too cute by half.

> if stepping function returned next then current version will be broken i think

I believe—though I'm not certain—that the new algorithm respects order. Whereas previously we used `Array#unshift` to add the elements of `$more` at the beginning of `$todo`, we now prepend the elements of `sx` (which are in reverse order having been consed) to `xs`. By consing twice we preserve the original order.

Your example required a bit of tweaking, but I believe this is what you wanted to test:

```javascript
> Z.chainRec (
  Array,
  (next, done, x) =>
    Array.from ({length: x.v},
                (_, k) => x.d ? done (k + 1) : next ({v: k + 1, d: true})),
  {v: 5, d: false}
)
[1, 1, 2, 1, 2, 3, 1, 2, 3, 4, 1, 2, 3, 4, 5]
```

Please give the code one more look. Now that I've explained the double consing you may be convinced that order is preserved. If not, let's add a few more test cases to be sure. :)

Reply:
Ah i see 👍